File: pdu/python/sensor_msgs/pdu_cdr_conv_PointCloud.py
from .pdu_pytype_PointCloud import PointCloud
from ..pdu_cdr_runtime import CdrReader, CdrWriter

from ..sensor_msgs.pdu_cdr_conv_ChannelFloat32 import *
from ..std_msgs.pdu_cdr_conv_Header import *
from ..geometry_msgs.pdu_cdr_conv_Point32 import *
from ..builtin_interfaces.pdu_cdr_conv_Time import *



def py_to_cdr_body_PointCloud(writer: CdrWriter, src: PointCloud):
    py_to_cdr_body_Header(writer, src.header)
    writer.write_sequence_length(src.points)
    for elem in src.points:
        py_to_cdr_body_Point32(writer, elem)
    writer.write_sequence_length(src.channels)
    for elem in src.channels:
        py_to_cdr_body_ChannelFloat32(writer, elem)


def cdr_body_to_py_PointCloud(reader: CdrReader, dst: PointCloud):
    cdr_body_to_py_Header(reader, dst.header)
    dst.points = []
    for _ in range(reader.read_uint32()):
        elem = Point32()
        cdr_body_to_py_Point32(reader, elem)
        dst.points.append(elem)
    dst.channels = []
    for _ in range(reader.read_uint32()):
        elem = ChannelFloat32()
        cdr_body_to_py_ChannelFloat32(reader, elem)
        dst.channels.append(elem)
    return dst


def py_to_cdr_PointCloud(src: PointCloud) -> bytes:
    writer = CdrWriter()
    writer.write_encapsulation()
    py_to_cdr_body_PointCloud(writer, src)
    return writer.bytes()


def cdr_to_py_PointCloud(cdr_payload) -> PointCloud:
    reader = CdrReader(cdr_payload)
    reader.read_encapsulation()
    dst = PointCloud()
    return cdr_body_to_py_PointCloud(reader, dst)

File: pdu/javascript/sensor_msgs/pdu_cdr_conv_PointCloud.js
import { PointCloud } from './pdu_jstype_PointCloud.js';
import { PduCdrWriter, PduCdrReader } from '../pdu_cdr_runtime.js';
import { ChannelFloat32 } from '../sensor_msgs/pdu_jstype_ChannelFloat32.js';
import { PduChannelFloat32Converter } from '../sensor_msgs/pdu_cdr_conv_ChannelFloat32.js';
import { Header } from '../std_msgs/pdu_jstype_Header.js';
import { PduHeaderConverter } from '../std_msgs/pdu_cdr_conv_Header.js';
import { Point32 } from '../geometry_msgs/pdu_jstype_Point32.js';
import { PduPoint32Converter } from '../geometry_msgs/pdu_cdr_conv_Point32.js';
import { Time } from '../builtin_interfaces/pdu_jstype_Time.js';
import { PduTimeConverter } from '../builtin_interfaces/pdu_cdr_conv_Time.js';


export class PduPointCloudConverter {
    /**
     * @param {PduCdrWriter} writer
     * @param { PointCloud } src
     */
    static to_cdr_body(writer, src) {
        PduHeaderConverter.to_cdr_body(writer, src.header);
        writer.write_sequence_length(src.points);
        for (const elem of src.points) {
            PduPoint32Converter.to_cdr_body(writer, elem);
        }
        writer.write_sequence_length(src.channels);
        for (const elem of src.channels) {
            PduChannelFloat32Converter.to_cdr_body(writer, elem);
        }
    }

    /**
     * @param {PduCdrReader} reader
     * @param { PointCloud } dst
     * @returns { PointCloud }
     */
    static cdr_body_to_js(reader, dst) {
        PduHeaderConverter.cdr_body_to_js(reader, dst.header);
        dst.points = [];
        for (let i = 0, len = reader.read_uint32(); i < len; i++) {
            dst.points.push(PduPoint32Converter.cdr_body_to_js(reader, new Point32()));
        }
        dst.channels = [];
        for (let i = 0, len = reader.read_uint32(); i < len; i++) {
            dst.channels.push(PduChannelFloat32Converter.cdr_body_to_js(reader, new ChannelFloat32()));
        }
        return dst;
    }

    /**
     * @param { PointCloud } src
     * @returns {ArrayBuffer}
     */
    static to_cdr(src) {
        const writer = new PduCdrWriter();
        writer.write_encapsulation();
        this.to_cdr_body(writer, src);
        return writer.get_buf();
    }

    /**
     * @param {ArrayBuffer|ArrayBufferView} cdrPayload
     * @returns { PointCloud }
     */
    static from_cdr(cdrPayload) {
        const reader = new PduCdrReader(cdrPayload);
        reader.read_encapsulation();
        return this.cdr_body_to_js(reader, new PointCloud());
    }
}
